### DISCUSSION TITLE
TVheadend and iptv-org's .m3u with ffmpeg MPEG-TS transcode

### DISCUSSION THREAD

**Discussion by HNJAMeindersma** [iptv-org/community#19] [Show and tell]
Running the plain `.m3u` files that iptv-org provides in TVheadend causes instability of the stream and depending on the installed version of TVheadend also of the system. This is caused because TVheadend is not used/made to handle a non MPEG-TS transport stream, which most/all of the streams in iptv-org's `.m3u`'s are not. This has nothing to do with iptv-org, but depends on the streams provider.

TVheadend [provides some documentation](https://tvheadend.org/projects/tvheadend/wiki/Custom_MPEG-TS_Input) on 'transcoding' the streams container to MPEG-TS. Generally there is no need to transcode the video or audio codec before the stream is ingested by TVheadend.

To automatically convert the URL's in iptv-org's `.m3u` files I've created a small shell script which is used together with a `pipe://` that you insert as an `Automatic IPTV Network` in the TVheadend webUI.

Create the following `piper.sh` file somewhere where the TVheadend user (usually called `hts`) can reach it:
```sh
#!/bin/sh
curl -sS $1 | sed '/^http/ s/$/ -vcodec copy -acodec copy -f mpegts -tune zerolatency pipe:1/' | sed '/^http/ s/^/pipe:\/\/\/usr\/bin\/ffmpeg -loglevel fatal -i /'
```
Then in the TVheadend webUI add an `Automatic IPTV Network` with in the URL-field: `pipe:///bin/sh /etc/tvheadend/piper.sh https://iptv-org.github.io/iptv/languages/eng.m3u`. The first part opens a pipe to a shell, the second part is the location of the above script. And at last you paste the iptv-org `.m3u` address.

Also, don't forget to set the `Service ID` to (usually) `1` for your `Automatic IPTV Network` to create all the services without scanning them. [More information here](https://tvheadend.org/projects/tvheadend/wiki/Automatic_IPTV_Network).

**Comment by BellezaEmporium** [comment]
Thank you for this useful advice. Pinned over the Show and Tell category.

**Comment by menotuu** [comment]
would this script made it possible to use plugins for input?
for example:  plugin://plugin.video.teliaplay-se?menu=play&streamType=live&streamId=1
and not https://.....

**Reply by HNJAMeindersma** [reply]
That might be possible. Since it's just a data stream with an MPEG format that ffmpeg picks up.
